rdt/transformers/categorical.py

```diff
@@ -455,13 +457,35 @@ def _fit(self, data):
     def _transform(self, data):
         """Replace each category with its corresponding integer value.
 
+        If a category has not been seen before, a random value is assigned.
+        If no categories have been fitted (and `data` is not empty), raise a `ValueError`.
+
         Args:
             data (pandas.Series):
                 Data to transform.
 
         Returns:
-            numpy.ndarray:
+            pd.Series.
         """
+        # Find the data categories that were not seen during fitting. Convert all nan's
+        # to np.nan to avoid having nan values with different ids.
+        unique_data = {np.nan if pd.isna(x) else x for x in pd.unique(data)}
+        categories = {np.nan if pd.isna(x) else x for x in self.categories_to_values.keys()}
+        unseen_categories = unique_data - categories
+        if unseen_categories:
+            if len(self.categories_to_values) == 0:
+                raise ValueError('No categories have been fitted.')
+
+            warn(
+                f'Warning: The data contains new categories {unseen_categories} that were not '
+                'seen in the original data. Assigning them random values. If you want to model '
+                'new categories, please fit the transformer again with the new data.'
+            )
+
+            for unseen_category in unseen_categories:
+                self.categories_to_values[unseen_category] = \
+                    np.random.randint(0, len(self.values_to_categories) - 1)
```

Comment:
Instead of doing this, you could pass a function into `map` that returns the value in the `self.categories_to_values` dictionary if it exists, and a random int if it doesn't. Something like
```
map_function = lambda x: self.categories_to_values.get(x, np.random.randint(0, len(self.values_to_categories) - 1))
return pd.Series(data).map(map_function)
```

Reply:
Hmm, we would not be returning the same value for each unseen category though, so if an unseen category shows up twice, each can get assigned a different value.